src/Sparrow.Server/Allocators.Fragment.cs

```diff
@@ -0,0 +1,272 @@
+                allocator.Allocated += segment.Size;
+
+                allocator._allocatedSegments.Add(segment);
+            }
+
+            allocator._currentBuffer = segment;
+            allocator._ptrCurrent = (byte*)segment.Address;
+
+            allocator._used = 0;            
+        }
+
+        public void Release(ref FragmentAllocator<TOptions> allocator, ref Pointer ptr)
+        {
+#if VALIDATION
+            // We dont want the internal composed allocator to check anything, if we messed, we messed here.
+            ptr.Generation = 0;
+#endif
+            
+            byte* address = (byte*)ptr.Address;
+            if (address < allocator._currentBuffer.Address || address != allocator._ptrCurrent - ptr.Size)
+            {
+                // We have fragmentation. note that this fragmentation will be healed by the call to Reset
+                // trying to do this on the fly is too expensive unless the chunk is big enough to consider it a whole segment for himself.
+                // Consider to compose this allocator with a PoolAllocator instead if high fragmentation is expected. 
+                if (ptr.Size > allocator._options.ReuseBlocksBiggerThan)
+                    allocator._internalPartialReusedSegments.Add(ptr.Size, ptr);
+            }
+            else
+            {
+                // since the returned allocation is at the end of the arena, we can just move
+                // the pointer back
+                allocator._used -= ptr.Size;
+                allocator._ptrCurrent -= ptr.Size;
+            }
+
+            allocator.InUse -= ptr.Size;
+            ptr = new Pointer();
+        }
+
+        public void Reset(ref FragmentAllocator<TOptions> allocator)
+        {
+            // These are all partial memory segments that cannot be released individually, only as part of the 
+            // whole segment; because if the underlying allocator does track the allocations they wont match.
+            allocator._internalPartialReusedSegments.Clear();
+
+            // Now we release the whole segments that the underlying allocator knows about.
+            foreach (var segment in allocator._allocatedSegments)
+            {
+                var ptr = segment;
+                allocator._internalAllocator.Release(ref ptr);
+            }
+
+            allocator._allocatedSegments.Clear();
+            allocator._currentBuffer = new Pointer();
+            allocator._ptrCurrent = null;
+            allocator._used = 0;
+
+            allocator.TotalAllocated = 0;
+            allocator.Allocated = 0;
+            allocator.InUse = 0;
+        }
+
+        public void OnAllocate(ref FragmentAllocator<TOptions> allocator, Pointer ptr) {}
+        public void OnRelease(ref FragmentAllocator<TOptions> allocator, Pointer ptr) {}
+
+        public void Dispose(ref FragmentAllocator<TOptions> allocator, bool disposing)
```

Comment:
`Reset` is identical to this, why?

Reply:
For all uses and purposes, we just can call reset. Check my earlier comments on the difference between `Reset` and `Renew`

Reply:
It was missing a call to the internal allocator `Dispose` method though... Fixed.